### PR TITLE
Accept `"{:}<"` in format macro parser

### LINKED ISSUE #112732: Format interprets closing brace as fill and suggesting to add another closing brace for no apperant reason.
### Code

```Rust
fn main() {
    println!("Hello, world! {0:}<3", 2);
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error: invalid format string: expected `'}'` but string was terminated
 --> src/main.rs:2:35
  |
2 |     println!("Hello, world! {0:}<3", 2);
  |                             -     ^ expected `'}'` in format string
  |                             |
  |                             because of this opening brace
  |
  = note: if you intended to print `{`, you can escape it using `{{`

error: could not compile `playground` (bin "playground") due to previous error
```

### Desired output

```Shell
The error should mention that the closing curly brace `}` is interpreted as the fill character and suggest the removal of the colon to fix the issue.
```

### Rationale and extra context

The format spec allows fill to be any character including `}`, this may be confusing.

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by TaKO8Ki**
I think this should also be compiled successfully.

Fixes #112732

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5a69e2326988b6150c62087ef60ecf2f

### CODE REVIEW

**Comment by compiler-errors** [comment]
For the record, this code would now break, right?

```rust
fn main() {
    println!("{:}<3}", 0);
}
```

**Comment by TaKO8Ki** [comment]
@compiler-errors Yes. After this change is applied, this case causes an error like the following, In the old parser, it is compiled successfully. I think this might be wrong behavior and new error looks better.

> According to the linked issue, the format spec allows any fill character, so shouldn't we be preserving this old behavior?

I guess we don't have detailed specification for formatting and the current behavior is regarded as a specification? Should I avoid changing the behavior of the format macro parse and just add a diagnostic for this case?

Given the following code:

```rust
fn main() {
    println!("{:}<3}", 0);
}
```

Before this change is applied: Compiled successfully

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=53f9ec33ce4fc8293d41b0fb061e0427

After the change is applied:
```
error: invalid format string: unmatched `}` found
 --> src/main.rs:2:20
  |
2 |     println!("{:}<3}", 0);
  |                    ^ unmatched `}` in format string
  |
  = note: if you intended to print `}`, you can escape it using `}}`

error: could not compile `debug_playground` (bin "debug_playground") due to previous error
```

**Comment by compiler-errors** [comment]
Nominating for T-libs-api I think, because this has to do with the specification of the format macro, and it's not strictly a fix.

File: tests/ui/fmt/format-args-issue-112732.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+// check-pass
+
+fn main() {
+    println!("Hello, world! {0:}<3", 2);
```
Comment (joshtriplett) [inline tests/ui/fmt/format-args-issue-112732.rs:4]: There's a missing close-brace here.

**Comment by TaKO8Ki** [comment]
@m-ou-se 

Thank you for discussing it. I see. The example you provided would cause an error like the following after this change is applied. Should `}` be available as the fill character?

```
error: invalid format string: unmatched `}` found
 --> src/main.rs:2:18
  |
2 |     format!("{:}>}", 2);
  |                  ^ unmatched `}` in format string
  |
  = note: if you intended to print `}`, you can escape it using `}}`

error: could not compile `debug_playground` (bin "debug_playground") due to previous error
```

> We should probably add that lint back in rustc, and that logic can then also help with giving a better diagnostic for "{:}>".

Ok. I'll create an issue in clippy and work on it when I have time.

**Comment by m-ou-se** [comment]
> Should `}` be available as the fill character?

Today, every single character is available as fill character. "Every single character except }" doesn't seem like a big improvement. (What about '{' or other potentially confusing ones?)

I agree `{:}>}` is confusing, but `{:}` is uncommon enough (and equivalent to `{}`) that I don't think it's worth coming up with an alternative syntax for specifying `}` as fill character.

**Comment by TaKO8Ki** [comment]
> Today, every single character is available as fill character. "Every single character except }" doesn't seem like a big improvement. (What about '{' or other potentially confusing ones?)

Understood. You're right. I'll close this pull request.

**Pull request closed without merging**